cub/cub/detail/rfa.cuh

```diff
@@ -0,0 +1,950 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#pragma once
+
+#include <cub/config.cuh>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__algorithm_>
+#include <cuda/std/array>
+#include <cuda/std/climits>
+#include <cuda/std/cmath>
+
+#ifndef __CUDACC__
+// disable zero checks
+#  define DISABLE_ZERO
+
+// disable nan / infinity checks
+#  define DISABLE_NANINF
+#endif
+
+// jump table for indexing into data
+#define _CUB_RFA_MAX_JUMP_ 5
+static_assert(_CUB_RFA_MAX_JUMP_ <= 5, "MAX_JUMP greater than max");
+
+CUB_NAMESPACE_BEGIN
+
+namespace detail
+{
+namespace rfa
+{
+
+template <typename Float, int LEN>
+static _CCCL_DEVICE Float* get_shared_bin_array()
+{
+  static __shared__ Float bin_computed_array[LEN];
+  return bin_computed_array;
+}
+
+// vector types with RFA are not supported currently
+#if 0
+template <class T>
+struct get_vector_type
+{};
+template <>
+struct get_vector_type<double>
+{
+  using type = double2;
+};
+template <>
+struct get_vector_type<float>
+{
+  using type = float4;
+};
+
+template <class T>
+using vector_t = typename get_vector_type<T>::type;
+
+template <class T>
+auto vector_size()
+{
+  return sizeof(vector_t<T>) / sizeof(T);
+}
+
+template <class T>
+_CCCL_FORCEINLINE static double abs_max(const T&);
+
+template <>
+_CCCL_DEVICE double abs_max(const float4& x)
+{
+  return fmax(fmaxf(fabs(x.x), fabs(x.y)), fmax(fabs(x.z), fabs(x.w)));
+}
+
+template <>
+_CCCL_DEVICE double abs_max(const double4& x)
+{
+  return fmax(fmax(fabs(x.x), fabs(x.y)), fmax(fabs(x.z), fabs(x.w)));
+}
+
+template <>
+_CCCL_DEVICE double abs_max(const double2& x)
+{
+  return fmax(fabs(x.x), fabs(x.y));
+}
+
+template <>
+_CCCL_DEVICE double abs_max(const float2& x)
+{
+  return fmax(fabs(x.x), fabs(x.y));
+}
+#endif
+
+template <class ftype>
+struct RFA_bins
+{
+  static constexpr int BIN_WIDTH = ::cuda::std::is_same_v<ftype, double> ? 40 : 13;
```

Comment:
We have recently  moved away from shouty `MACRO_NAMES`, could we rename this and others to `bin_width`?
@bernhardmgruber because I am not 100% sure whether that is consensus on the CUB side

Reply:
CUB uses `snake_case` everywhere, except for template parameters which have `PascalCase`.